examples/vision/attention_mil_classification.py

```diff
@@ -0,0 +1,575 @@
+The goal here is to be able to know which key chain (in general) and key (in specific) can
+open up the locked door.
+
+In order to solve this task, we shall find the key that can opens the door. If the key was
+correctly found, then the entire key chain will be classified as positive. If the key chain
+doesn't contain any compatible key, it will be classified as negative.
+
+In the context of MIL, each of the key chains are known as **bags**, in which each can be
+labelled as positive or negative. Each key within the key chains are known as **instances**.
+
+## Introduction
+
+Usually for image classification tasks using **supervised learning**  algorithms, the learner
+receives a single label for a single image. In case of **MIL**, the learner receives a single
+label for a set of images.
+
+The motivation behind this model is to:
+
+- know which instance (in this case: image) has contributed the most to the class label
+prediction.
+- leverage from weakly labelled samples.
+
+### MIL Classifier: Attention-based approach
+
+The MIL classifier is modelled using neural networks. The attention mechanism as MIL pooling:
+
+- Trainable MIL operator.
+- Interpretable results using attention scores.
+
+Using attention scores or weights could be very useful for results interpretability as it will
+not merely predict the final output class but the instance scores that led to those
+classification results.
+
+Another reason why this method could be helpful is: weakly labelled data. Since labelling
+individual samples is often performed manually, it is considered to be a time consuming task.
+This task is also prone to errors (especially in computer vision problems).
+
+Therefore, instead of having individual labelled data points or instances, this algorithm
+receives a single label for each bag of instances. The bag is labelled positive if it contains
+atleast one positive instance and negative if it does not contain any.
+
+The attention mechanism provides insight into the contribution of each instance to the bag label.
+
+Besides the implementation of this mechanism, some **regularization** techniques,
+**ensemble averaging** (for stability) and dealing with **imbalanced data** will be covered.
+
+## References
+
+- [Attention-based Deep Multiple Instance Learning](https://arxiv.org/pdf/1802.04712.pdf).
+- Some of attention operator code implementation was inspired from https://github.com/utayao/Atten_Deep_MIL.
+- Imbalanced data [tutorial](https://www.tensorflow.org/tutorials/structured_data/imbalanced_data)
+by TensorFlow.
+"""
+"""
+## Setup
+"""
+
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+from tqdm import tqdm
+from matplotlib import pyplot as plt
+
+plt.style.use("ggplot")
+
+"""
+## Create dataset
+`mnist` and `fashion_mnist` datasets have been tested using this implementation. In this
+example, I will be using `mnist`.
```

Comment:
This is not the standard usage of MNIST, it would be useful to provide an illustration of how the data is used.

Reply:
That's right. I did provide now a short illustration.